direct/transport/tpm.go

```diff
@@ -10,20 +10,72 @@ import (
 // TPM represents a logical connection to a TPM.
 type TPM interface {
 	Send(input []byte) ([]byte, error)
-	Close() error
 }
 
-// LocalTPM represents a connection to the local TPM.
-type LocalTPM struct {
+// TPMCloser represents a logical connection to a TPM and you can close it.
+type TPMCloser interface {
+	TPM
+	io.Closer
+}
+
+
+// wrappedRW represents a struct that wraps an io.ReadWriter
+// to a transport.TPM to be compatible with tpmdirect.
+type wrappedRW struct {
+	transport io.ReadWriter
+}
+
+// wrappedRWC represents a struct that wraps an io.ReadWriteCloser
+// to a transport.TPM to be compatible with tpmdirect.
+type wrappedRWC struct {
 	transport io.ReadWriteCloser
 }
 
+// wrappedTPM represents a struct that wraps a transport.TPM's underlying 
+// transport to use with legacy code that expects an io.ReadWriter.
+type wrappedTPM struct {
+	response []byte
+	tpm      TPM
+}
+
+// FromReadWriter takes in a io.ReadWriter and returns a
+// transport.TPM wrapping the io.ReadWriter.
+func FromReadWriter(rw io.ReadWriter) TPM {
+	return &wrappedRW{transport: rw}
+}
+
+// ToReadWriter takes in a transport TPM and returns an
+// io.ReadWriter wrapping the transport TPM.
+func ToReadWriter(tpm TPM) io.ReadWriter {
+	return &wrappedTPM{tpm: tpm}
+}
+
+// Read copies t.response into the p buffer and return the appropriate length.
+func (t *wrappedTPM) Read(p []byte) (n int, err error) {
+	p = t.response
```

Comment:
So I don't think this works, using this implementation we won't actually write any bytes to p. This just overwrites p temporarily in the function, but doesn't actually do anything. Example:
https://go.dev/play/p/Lqdl8wT_1n1

I think we will need to either use a `*bytes.Buffer` as the internal response type (example: https://go.dev/play/p/JW4V94-0Blu), or have an implementation that looks like:
```go
func (t *wrappedTPM) Read(p []byte) (int, error) {
	n := copy(p, t.response)
	t.response = t.response[n:]
	if len(t.response) == 0 {
		return n, io.EOF
	}
	return n, nil
}
```
see this example: https://go.dev/play/p/Hppzq-rQETY
which uses [the `copy` builtin](https://pkg.go.dev/builtin@go1.18.3#copy)

Reply:
Fixed, should work as intended now.